File: constellation_planning/algorithms/tabu_search.py
# -*- coding: utf-8 -*-
"""禁忌搜索算法"""

import random
from typing import List, Any, Tuple
from collections import deque
from .base import PlanningAlgorithm, Solution, AlgorithmConfig


class TabuSearch(PlanningAlgorithm):
    """禁忌搜索算法"""
    
    def __init__(
        self, 
        config: AlgorithmConfig = None,
        tabu_tenure: int = 10,
        neighborhood_size: int = 20
    ):
        super().__init__(config)
        self.tabu_tenure = tabu_tenure
        self.neighborhood_size = neighborhood_size
        self.tabu_list = deque(maxlen=tabu_tenure)
    
    def get_name(self) -> str:
        return f"TabuSearch(tenure={self.tabu_tenure})"
    
    def solve(
        self,
        observations: List[Any],
        satellites: List[Any],
        **kwargs
    ) -> Solution:
        """求解任务规划问题"""
        self._start_timer()
        
        if self.config.random_seed:
            random.seed(self.config.random_seed)
        
        obs_list = list(observations)
        # 建立 ID 到 分数的映射
        self.score_map = {
            obs.id: getattr(obs, 'score', getattr(obs, 'priority', 1.0)) 
            for obs in obs_list
        }
        
        # 1. 生成初始解
        current = self._generate_initial_solution(obs_list, satellites)
        self._update_best(current)
        
        # 2. 迭代搜索
        for iteration in range(self.config.max_iterations):
            if self._is_time_exceeded():
                break
            
            # 生成邻域
            neighbors = self._generate_neighborhood(current, obs_list, satellites)
            
            # 选择最佳非禁忌移动（或满足特赦条件）
            best_neighbor = None
            best_move = None
            
            for neighbor, move in neighbors:
                # 检查禁忌
                if move in self.tabu_list:
                    # 特赦条件：比历史最优更好
                    if neighbor.objective_value > self.best_solution.objective_value:
                        if best_neighbor is None or neighbor.objective_value > best_neighbor.objective_value:
                            best_neighbor = neighbor
                            best_move = move
                else:
                    if best_neighbor is None or neighbor.objective_value > best_neighbor.objective_value:
                        best_neighbor = neighbor
                        best_move = move
            
            if best_neighbor:
                current = best_neighbor
                self.tabu_list.append(best_move)
                self._update_best(current)
            
            self.history.append(self.best_solution.objective_value)
        
        return self.best_solution
    
    def _calculate_score(self, solution: Solution) -> float:
        """计算解的总分"""
        return sum(self.score_map.get(obs_id, 1.0) for obs_id in solution.assignments)
    
    def _generate_initial_solution(
        self,
        observations: List[Any],
        satellites: List[Any]
    ) -> Solution:
        """生成初始解（贪婪法）"""
        solution = Solution()
        
        # 简单贪婪：按优先级分配
        sorted_obs = sorted(observations, key=lambda x: self.score_map.get(x.id, 1.0), reverse=True)
        
        for obs in sorted_obs:
            # 分配给第一个可用的卫星
            if hasattr(obs, 'satellite_id'):
                solution.assignments[obs.id] = obs.satellite_id
        
        solution.objective_value = self._calculate_score(solution)
        return solution
    
    def _generate_neighborhood(
        self,
        current: Solution,
        observations: List[Any],
        satellites: List[Any]
    ) -> List[Tuple[Solution, str]]:
        """生成邻域解"""
        neighbors = []
        obs_list = list(observations)
        
        for _ in range(min(self.neighborhood_size, len(obs_list))):
            neighbor = current.copy()
            
            # 随机选择一个观测进行调整
            obs = random.choice(obs_list)
            move = f"flip_{obs.id}"
            
            if obs.id in neighbor.assignments:
                # 移除分配
                del neighbor.assignments[obs.id]
            else:
                # 添加分配
                if hasattr(obs, 'satellite_id'):
                    neighbor.assignments[obs.id] = obs.satellite_id
            
            neighbor.objective_value = self._calculate_score(neighbor)
            neighbors.append((neighbor, move))
        
        return neighbors

File: matlab_version/+algorithms/TabuSearch.m
classdef TabuSearch <
    algorithms.PlanningAlgorithm %
        TABUSEARCH 禁忌搜索 MATLAB 实现

            properties TabuTenure NeighborhoodSize TabuList end

                methods function obj = TabuSearch(config) obj =
    obj @algorithms.PlanningAlgorithm(config);
obj.TabuTenure = 10;
obj.NeighborhoodSize = 20;
obj.TabuList = {};
% Cell array of strings

    if isfield (config, 'TabuTenure'),
    obj.TabuTenure = config.TabuTenure;
end end

    function bestSol = solve(obj, observations, satellites) tic;

% 1. 预处理 obsMap = containers.Map();
            for
              i = 1 : length(observations),
              obsMap(observations(i).Id) = observations(i);
            end obsIds = obsMap.keys;

            % 2. 初始解(贪婪) current = obj.generateGreedySolution(observations,
                                                                   satellites);
            currentScore = obj.evaluate(current, obsMap);

            obj.BestSolution = current;
            obj.BestSolution.Objective = currentScore;

            % 3. 迭代
            for iter = 1:obj.Config.MaxIterations
                if toc > obj.Config.TimeLimit, break;
            end

                % 生成邻域 bestNeighbor = [];
            bestNeighborScore = -Inf;
            bestMove = '';

                for
                  k = 1 : obj.NeighborhoodSize % 随机翻转 obsId =
                          obsIds{randi(length(obsIds))};
                neighbor = current;
                % struct copy by value ? No,
                    assignments map is reference ! %
                        Map is handle class.So we must copy map
                            explicitly.neighbor.Assignments = containers.Map(
                        current.Assignments.keys, current.Assignments.values);

                move = ['flip_' obsId];
                if neighbor
                  .Assignments.isKey(obsId) remove(neighbor.Assignments, obsId);
                else
                  % 简化：如果不在解中，尝试加入（假设 SatID 固定） obs =
                      obsMap(obsId);
                neighbor.Assignments(obsId) = obs.SatelliteId;
                end

                    score = obj.evaluate(neighbor, obsMap);

                % 检查禁忌 isTabu = any(strcmp(obj.TabuList, move));
                isAspiration = (score > obj.BestSolution.Objective);

                if
                  ~isTabu || isAspiration if score >
                                 bestNeighborScore bestNeighbor = neighbor;
                bestNeighborScore = score;
                bestMove = move;
                end end end

                    %
                    更新当前解 if ~isempty(bestNeighbor) current = bestNeighbor;
                currentScore = bestNeighborScore;

                % 更新禁忌表 obj.TabuList{end + 1} = bestMove;
                if length (obj.TabuList)
                  > obj.TabuTenure obj.TabuList(1) = [];
                end

                        % 更新全局最优 if currentScore >
                    obj.BestSolution.Objective obj.BestSolution = current;
                obj.BestSolution.Objective = currentScore;
                end end

                    obj.History(end + 1) = obj.BestSolution.Objective;
                end

                    obj.Runtime = toc;
                bestSol = obj.BestSolution;
                end end

                    methods(Access = private) function sol =
                        generateGreedySolution(obj, observations, satellites)
                            sol.Assignments = containers.Map();
                % 简单按分数排序[~, idx] =
                    sort([observations.Score], 'descend');
            for
              k = 1 : length(idx) obs = observations(idx(k));
            % 贪婪加入 sol.Assignments(obs.Id) = obs.SatelliteId;
            end sol.Objective = 0;
            end

                function score = evaluate(obj, solution, obsMap) score = 0;
            keys = solution.Assignments.keys;
            for
              i = 1 : length(keys) if obsMap.isKey(keys{i}) score =
                  score + obsMap(keys{i}).Score;
            end end end end end
